server/server.go

```diff
@@ -142,8 +124,32 @@ func scheme(r *http.Request) string {
 
 // URLRoot builds a string containing the scheme, host and port based on a combination of user defined values,
 // headers and request parameters. The function is public so it can be overridden for other implementations.
-var URLRoot = func(r *http.Request) string {
-	return fmt.Sprintf("%v://%v", scheme(r), hostName(r))
+var URLRoot = func(r *http.Request) *url.URL {
+	root := url.URL{
+		Scheme: scheme(r),
+		Host:   hostName(r),
+	}
+
+	return &root
+}
+
+// buildCapabilitiesURL is responsible for building the various URLs which are returned by
+// the capabilities endpoints using the request, uri parts, and query params the function
+// will determine the protocol host:port and URI prefix that need to be included based on
+// user defined configurations and request context
+func buildCapabilitiesURL(r *http.Request, uriParts []string, query url.Values) string {
+	uri := path.Join(uriParts...)
+	q := query.Encode()
+	if q != "" {
+		// prepend our query identifier
+		q = "?" + q
+	}
+
+	// usually the url.URL package would be used for building the URL, but the
+	// uri template for the tiles contains characters that the package does not like:
+	// {z}/{x}/{y}. These values are escaped during the String() call which does not
+	// work for the capabilities URLs.
+	return fmt.Sprintf("%v%v%v", URLRoot(r), path.Join(URIPrefix, uri), q)
```

Comment:
here you can just do `path.Join(URIPrefix, uriParts...)`

Reply:
path.Join() is variadic and does not like that syntax. I tried it before I ended up with this code. Check it out: https://play.golang.org/p/fSUEykW6Je3

Reply:
My bad I thought it was `func (string, string...) string`